stan/math/prim/fun/log_mix.hpp

```diff
@@ -245,142 +161,37 @@ return_type_t<T_theta, std::vector<Eigen::Matrix<T_lam, R, C>>> log_mix(
     check_consistent_sizes(function, "theta", theta, "lambda", lambda[n]);
   }
 
-  scalar_seq_view<T_theta> theta_vec(theta);
-  T_partials_vec theta_dbl(M);
-  for (int m = 0; m < M; ++m) {
-    theta_dbl[m] = value_of(theta_vec[m]);
-  }
+  const auto& theta_dbl = value_of(as_column_vector_or_scalar(theta));
 
   T_partials_mat lam_dbl(M, N);
-  vector_seq_view<T_lamvec_type> lam_vec(lambda);
   for (int n = 0; n < N; ++n) {
-    for (int m = 0; m < M; ++m) {
-      lam_dbl(m, n) = value_of(lam_vec[n][m]);
-    }
+    lam_dbl.col(n) = value_of(as_column_vector_or_scalar(lambda[n]));
   }
 
   T_partials_mat logp_tmp = log(theta_dbl).replicate(1, N) + lam_dbl;
-
   T_partials_vec logp(N);
   for (int n = 0; n < N; ++n) {
-    logp[n] = log_sum_exp(logp_tmp.col(n).eval());
+    logp[n] = log_sum_exp(logp_tmp.col(n));
```

Comment:
Looking forward to 3.4, which will introduce [iterators over rows/columns](https://eigen.tuxfamily.org/dox-devel/group__TutorialSTL.html) so we don't need the ```logp_tmp``` intermediary

Reply:
As I see it we would need to implement `log_sum_exp` that would work colwise to eliminate the temporary.